runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/expression/ExpressionResolver.java

```diff
@@ -0,0 +1,59 @@
+package io.aklivity.zilla.runtime.engine.expression;
+
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Objects.requireNonNull;
+import static java.util.ServiceLoader.load;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.regex.Pattern;
+
+public final class ExpressionResolver
+{
+    private static final Pattern EXPRESSION_PATTERN =
+            Pattern.compile("\\$\\{\\{\\s*([^\\s\\}]*)\\.([^\\s\\}]*)\\s*\\}\\}");
+
+    private final Map<String, ExpressionResolverSpi> resolverSpis;
+
+    public static ExpressionResolver instantiate()
+    {
+        return instantiate(load(ExpressionResolverSpi.class));
+    }
+
+    public String resolve(
+        String config)
+    {
+        return EXPRESSION_PATTERN.matcher(config)
```

Comment:
Instead of recreating the `Matcher` on every call to `resolve(config)`, suggest instead creating the matcher via `EXPRESSION_MATCHER.matcher("")` in constructor, then `matcher.reset(config).replaceAll(...)` in `resolve` method.

Also suggest changing parameter names from `config` to `template`, since config is just one possible usage of this expression resolver.